### PR TITLE
Fixed #34856 -- Fixed references to index_together in historical migrations.

### PR DESCRIPTION
**Pull request by charettes**
#### Trac ticket number

ticket-34856

Follow up to [this forum discussion](https://forum.djangoproject.com/t/django-5-1-alterindextogether-raising-typeerror-class-meta-got-invalid-attribute-s-index-together/36450) about the broken guarantee that 

> AlterIndexTogether is officially supported only for pre-Django 4.2 migration files. For backward compatibility reasons, it’s still part of the public API, and there’s no plan to deprecate or remove it, but it should not be used for new migrations. Use AddIndex and RemoveIndex operations instead.

Thinking more about it I believe this is a good solution to keeping minimal support for historical `AlterIndexTogether`. I do wonder if we should restore some of the tests removed in 2abf417c815c20f41c0868d6f66520b32347106e to ensure it still works properly though or maybe the one added here is sufficient?

### CODE REVIEW

**Comment by sarahboyce** [comment]
Thank you for doing this @charettes 
I was having a look at our test coverage for index together, and since the test removals we are missing [some coverage here](https://djangoci.com/job/django-coverage/HTML_20Coverage_20Report/z_4a66af0e9ba9af61_autodetector_py.html#t1360)

Adding one test back to the autodetector adds this coverage back, I think I would suggest we do this:
```diff
--- a/tests/migrations/test_autodetector.py
+++ b/tests/migrations/test_autodetector.py
@@ -5148,6 +5148,35 @@ class AutodetectorTests(BaseAutodetectorTests):
             preserve_default=True,
         )
 
+    def test_rename_historic_index_together_to_index(self):
+        book_index_together = ModelState(
+            "otherapp",
+            "Book",
+            [
+                ("id", models.AutoField(primary_key=True)),
+                ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
+                ("title", models.CharField(max_length=200)),
+            ],
+            {
+                "index_together": {("author", "title")},
+            },
+        )
+        changes = self.get_changes(
+            [AutodetectorTests.author_empty, book_index_together],
+            [AutodetectorTests.author_empty, AutodetectorTests.book_indexes],
+        )
+        self.assertNumberMigrations(changes, "otherapp", 1)
+        self.assertOperationTypes(changes, "otherapp", 0, ["RenameIndex"])
+        self.assertOperationAttributes(
+            changes,
+            "otherapp",
+            0,
+            0,
+            model_name="book",
+            new_name="book_title_author_idx",
+            old_fields=("author", "title"),
+        )
+
 
 class MigrationSuggestNameTests(SimpleTestCase):
```

<details><summary>We could in theory add all of these but I don't think it's necessary</summary>

```python
class AutodetectorHistoricIndexTogetherTests(BaseAutodetectorTests):
    book_index_together = ModelState(
        "otherapp",
        "Book",
        [
            ("id", models.AutoField(primary_key=True)),
            ("author", models.ForeignKey("testapp.Author", models.CASCADE)),
            ("title", models.CharField(max_length=200)),
        ],
        {
            "index_together": {("author", "title")},
        },
    )

    def test_empty_index_together(self):
        """Empty index_together shouldn't generate a migration."""
        # Explicitly testing for not specified, since this is the case after
        # a CreateModel operation w/o any definition on the original model
        model_state_not_specified = ModelState(
            "a", "model", [("id", models.AutoField(primary_key=True))]
        )
        # Explicitly testing for None, since this was the issue in #23452 after
        # an AlterIndexTogether operation with e.g. () as value
        model_state_none = ModelState(
            "a",
            "model",
            [("id", models.AutoField(primary_key=True))],
            {
                "index_together": None,
            },
        )
        # Explicitly testing for the empty set, since we now always have sets.
        # During removal (('col1', 'col2'),) --> () this becomes set([])
        model_state_empty = ModelState(
            "a",
            "model",
            [("id", models.AutoField(primary_key=True))],
            {
                "index_together": set(),
            },
        )

        def test(from_state, to_state, msg):
            changes = self.get_changes([from_state], [to_state])
            if changes:
                ops = ", ".join(
                    o.__class__.__name__ for o in changes["a"][0].operations
                )
                self.fail("Created operation(s) %s from %s" % (ops, msg))

        tests = (
            (
                model_state_not_specified,
                model_state_not_specified,
                '"not specified" to "not specified"',
            ),
            (model_state_not_specified, model_state_none, '"not specified" to "None"'),
            (
                model_state_not_specified,
                model_state_empty,
                '"not specified" to "empty"',
            ),
            (model_state_none, model_state_not_specified, '"None" to "not specified"'),
            (model_state_none, model_state_none, '"None" to "None"'),
            (model_state_none, model_state_empty, '"None" to "empty"'),
            (
                model_state_empty,
                model_state_not_specified,
                '"empty" to "not specified"',
            ),
            (model_state_empty, model_state_none, '"empty" to "None"'),
            (model_state_empty, model_state_empty, '"empty" to "empty"'),
        )

        for t in tests:
            test(*t)

    def test_rename_index_together_to_index(self):
        changes = self.get_changes(
            [AutodetectorTests.author_empty, self.book_index_together],
            [AutodetectorTests.author_empty, AutodetectorTests.book_indexes],
        )
        self.assertNumberMigrations(changes, "otherapp", 1)
        self.assertOperationTypes(changes, "otherapp", 0, ["RenameIndex"])
        self.assertOperationAttributes(
            changes,
            "otherapp",
            0,
            0,
            model_name="book",
            new_name="book_title_author_idx",
            old_fields=("author", "title"),
        )

    def test_index_together_no_changes(self):
        """
        index_together doesn't generate a migration if no changes have been
        made.
        """
        changes = self.get_changes(
            [AutodetectorTests.author_empty, self.book_index_together],
            [AutodetectorTests.author_empty, self.book_index_together],
        )
        self.assertEqual(len(changes), 0)

    def test_add_model_order_with_respect_to_index_together(self):
        changes = self.get_changes(
            [],
            [
                AutodetectorTests.book,
                ModelState(
                    "testapp",
                    "Author",
                    [
                        ("id", models.AutoField(primary_key=True)),
                        ("name", models.CharField(max_length=200)),
                        ("book", models.ForeignKey("otherapp.Book", models.CASCADE)),
                    ],
                    options={
                        "order_with_respect_to": "book",
                        "index_together": {("name", "_order")},
                    },
                ),
            ],
        )
        self.assertNumberMigrations(changes, "testapp", 1)
        self.assertOperationTypes(changes, "testapp", 0, ["CreateModel"])
        self.assertOperationAttributes(
            changes,
            "testapp",
            0,
            0,
            name="Author",
            options={
                "order_with_respect_to": "book",
                "index_together": {("name", "_order")},
            },
        )
```
</details>

**Comment by charettes** [comment]
@sarahboyce thanks for skimming through the tests! I think that adding back `test_rename_historic_index_together_to_index` only makes sense.

**Comment by sarahboyce** [comment]
@andres-holvi @laymonage can you test this patch against your projects and confirm whether this solves the issue for you?

**Review by laymonage — APPROVED** [review]
I'm not quite familiar with the migrations internals to comment on the fix, but the test makes sense and I can confirm [it does fix the reproducible example](https://github.com/laymonage/django-test-repro/actions/runs/12297622294/job/34319180032) I made last year 😄

**Comment by systemallica** [comment]
Hi! This is @andres-holvi from another account. Sorry, it seems I'm late to the party 😄 

Thanks a lot for looking into this! Yes I can confirm the changes here solve the issue I was having. 

Just for my education: I have added back [here](https://github.com/django/django/pull/18931) some of the tests removed in https://github.com/django/django/commit/2abf417c815c20f41c0868d6f66520b32347106e and test_rename_field_index_together specifically required reverting some changes to `django.db.migrations.state.rename_field`. Could anyone confirm if this is something relevant, or something that is not applicable anymore? Sorry, I am not too familiar with the big picture here.

**Comment by charettes** [comment]
@systemallica

> Could anyone confirm if this is something relevant, or something that is not applicable anymore? Sorry, I am not too familiar with the big picture here.

I believe they are yes as historical `AlterIndexTogether` are not going away unless manually removed.

If we are going to re-instate more tests as you did in https://github.com/django/django/pull/18931 (thank you for doing that btw) I think we should drop `test_historical_index_together` and use `test_alter_index_together` as to not confuse future readers.

TL;DR +1 to add the `ProjectState.rename_field` bits of logic and include back all the `test_operations` tests like you did with their original names.

**Comment by systemallica** [comment]
@charettes Thanks for checking. I have just f-pushed my branch with those changes + added release notes to backport this to 5.1.5, as mentioned in the [docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#bugs).

Let me know if anything else is needed!

**Comment by sarahboyce** [comment]
Superseded by https://github.com/django/django/pull/18931
